src/commands/publish.rs

```diff
@@ -162,38 +235,154 @@ pub fn publish(publish_opts: PublishOpt) -> anyhow::Result<()> {
+    }
+
+    let get_google_signed_url = GetSignedUrl::build_query(get_signed_url::Variables {
         name: package.name.to_string(),
         version: package.version.to_string(),
-        description: package.description.clone(),
-        manifest: manifest_string,
-        license: package.license.clone(),
-        license_file,
-        readme,
-        repository: package.repository.clone(),
-        homepage: package.homepage.clone(),
-        file_name: Some(archive_name.clone()),
-        signature: maybe_signature_data,
     });
-    assert!(archive_path.exists());
-    assert!(archive_path.is_file());
 
-    if !publish_opts.dry_run {
-        let _response: publish_package_mutation::ResponseData =
-            execute_query_modifier(&q, |f| f.file(archive_name, archive_path).unwrap())
-                .map_err(on_error)?;
+    let _response: get_signed_url::ResponseData =
+        execute_query_modifier(&get_google_signed_url, |f| f)?;
+
+    let url = _response.url.ok_or_else(|| {
+        anyhow!(
+            "could not get signed url for package {}@{}",
+            package.name,
+            package.version
+        )
+    })?;
+
+    let signed_url = url.url;
+    let url = url::Url::parse(&signed_url).unwrap();
+    let client = reqwest::blocking::Client::builder()
+        .default_headers(reqwest::header::HeaderMap::default())
+        .build()
+        .unwrap();
+
+    let res = client
+        .post(url)
+        .header(reqwest::header::CONTENT_LENGTH, "0")
+        .header(reqwest::header::CONTENT_TYPE, "application/octet-stream")
+        .header("x-goog-resumable", "start");
+
+    let result = res.send().unwrap();
+
+    if result.status() != reqwest::StatusCode::from_u16(201).unwrap() {
+        return Err(anyhow!(
+            "Uploading package failed: got HTTP {:?} when uploading",
+            result.status()
+        ));
+    }
+
+    let headers = result
+        .headers()
+        .into_iter()
+        .filter_map(|(k, v)| {
+            let k = k.to_string();
+            let v = v.to_str().ok()?.to_string();
+            Some((k.to_lowercase(), v))
+        })
+        .collect::<BTreeMap<_, _>>();
+
+    let session_uri = headers.get("location").unwrap().clone();
+
+    let total = archived_data_size;
+
+    use indicatif::{ProgressBar, ProgressState, ProgressStyle};
+
+    // archive_path
+    let mut file = std::fs::OpenOptions::new()
+        .read(true)
+        .open(archive_path)
+        .map_err(|e| anyhow!("cannot open archive {}: {e}", archive_path.display()))?;
+
+    let pb = ProgressBar::new(archived_data_size);
+    pb.set_style(ProgressStyle::with_template("{spinner:.green} [{elapsed_precise}] [{wide_bar:.cyan/blue}] {bytes}/{total_bytes} ({eta})")
+    .unwrap()
+    .with_key("eta", |state: &ProgressState, w: &mut dyn Write| {
+        write!(w, "{:.1}s", state.eta().as_secs_f64()).unwrap()
+    })
+    .progress_chars("#>-"));
+
+    let chunk_size = 256 * 1024;
```

Comment:
From your experience, is there a particular upload size that will lead to time outs? 

It's about 200ms from Sydney to New York (one way), so if our chunks are too small we'll spend a lot of time just waiting for messages to cross the pacific.

Reply:
tested: 1MB is best (so that the user can see progress): ~315s / 100MB versus 312s when using 2MB blocks: the extra 3 seconds aren't really worth it, it's more important to show progress during uploading